language-plutus-core/src/Language/PlutusCore/TypeCheck/Internal.hs

```diff
@@ -303,17 +305,105 @@ unfoldFixOf pat arg k = do
 -- a static built-in name is allowed to depend on a dynamic built-in type which are not required
 -- to be normalized. For dynamic built-in names we store a map from them to their *normalized types*,
 -- with the normalization happening in this module, but what should we do for static built-in names?
 -- Right now we just renormalize the type of a static built-in name each time we encounter that name.
-inferTypeOfBuiltinM (BuiltinName    _   name) = normalizeType $ typeOfBuiltinName name
+inferTypeOfBuiltinM _ (StaticBuiltinName name) = normalizeType $ typeOfBuiltinName name
 -- TODO: inline this definition once we have only dynamic built-in names.
-inferTypeOfBuiltinM (DynBuiltinName ann name) = lookupDynamicBuiltinNameM ann name
+inferTypeOfBuiltinM ann (DynBuiltinName name)  = lookupDynamicBuiltinNameM ann name
+
+
+-- | Peel the type parameters off the front of the type of a builtin,
+-- building a mapping from type parameters to the types they're being
+-- instantiated at (these are included in the ApplyBuiltin
+-- constructor).  When we get to the end, carry on and look at the
+-- term arguments of the application.
+checkBuiltinTypeArgs
+    :: (GShow uni, GEq uni, DefaultUni <: uni)
+    => [(TyName, Normalized (Type TyName uni ()))]
+    -> BuiltinName
+    -> Type TyName uni ()
+    -> [Type TyName uni ann]
+    -> [Value TyName Name uni ann]
+    -> TypeCheckM uni ann (Normalized (Type TyName uni ()))
+checkBuiltinTypeArgs mapping bn (TyForall () tyname kind ty') (tyarg:tyargs) args = do
+  checkKindM  (typeAnn tyarg) tyarg (void kind)
+  tyarg' <- normalizeType (void tyarg)
+  checkBuiltinTypeArgs ((tyname,tyarg'):mapping) bn ty' tyargs args
+checkBuiltinTypeArgs _       bn (TyForall _ _tyname _kind _ty') [] _ = Prelude.error $ "Underinstantiated: " ++ show bn
+checkBuiltinTypeArgs mapping bn ty' [] args = checkBuiltinTermArgs mapping bn ty' args
+checkBuiltinTypeArgs _       bn _  (_:_) _ = Prelude.error $ "Overinstantiated: " ++ show bn -- We've got a forall type but nowhere to instantiate it
+-- FIXME: proper errors
+
+
+-- PROBLEM.  Suppose we have a builtin application (bn tys args). Can
+-- we tell if it's underapplied?  Suppose the Type of bn is int -> int
+-- -> int.  Ho do we know whether it takes two arguments or one (or
+-- even zero)?  It could take two integers and return another, or take
+-- one integer and return a function.  The type scheme for the builtin
+-- contains this information, but I don't think we have access to it
+-- here.  The problem is that different type schemes can translate to
+-- the same Type, losing information on the way.  We can solve this
+-- for the time being by checking that the type we're left with at the
+-- end isn't functional, which may not work for polymorphic builtins,
+-- eg  o : forall a b c: (b->c) -> (a->b) -> (a->c).
+-- Things would be a lot easier if (a) we had the type schema for the
+-- builtin at this point or (b) builtin names carried their arities
+-- about with them.
```

Comment:
You have the type scheme of a builtin at any point. For a dynamic builtin it amounts to performing a lookup in the environment. And for a static builtin you can call [`withTypedBuiltinName`](https://github.com/input-output-hk/plutus/blob/25eea28df681f74fd0ce4e8e483b619e52a38220/language-plutus-core/src/Language/PlutusCore/Constant/Name.hs#L60). That should change the way you do type checking.